### PR TITLE
Fix 5315: use Kokkos::atomic_load to Correct Race Condition Giving Rise to Seg Fault'ing Error in OpenMP tests 

### PR DESCRIPTION
**Pull request by ajpowelsnl**
This fix addresses seg fault failures surfaced in the `KokkosCore_UnitTest_OpenMP` (`openmp.task_fib_single`) in issue #5315.  Changes were made in `Kokkos_WorkGraphPolicy.hpp`.

### CODE REVIEW

**Comment by crtrott** [comment]
Also removed all volatile attributes, since the pointers are now exclusively accessed through atomic functions.

File: core/src/Kokkos_WorkGraphPolicy.hpp [hunk 1]
```diff
@@ -130,14 +130,14 @@ class WorkGraphPolicy : public Kokkos::Impl::PolicyTraits<Properties...> {
...
+    std::int32_t* const ready_queue = &m_queue[0];
+    std::int32_t* const begin_hint  = &m_queue[2 * N];
 
     // begin hint is guaranteed to be less than or equal to
     // actual begin location in the queue.
 
-    for (std::int32_t i = *begin_hint; i < N; ++i) {
-      const std::int32_t w = ready_queue[i];
+    for (std::int32_t i = Kokkos::atomic_load(begin_hint); i < N; ++i) {
+      const std::int32_t w = Kokkos::atomic_load(&ready_queue[+i]);
```
Comment (dalg24) [inline core/src/Kokkos_WorkGraphPolicy.hpp:140]:
```suggestion
      const std::int32_t w = Kokkos::atomic_load(&ready_queue[i]);
```

**Comment by PhilMiller** [comment]
Could the variables in question be declared as `std::atomic`? This is strictly host code, right?

**Comment by ajpowelsnl** [comment]
> Could the variables in question be declared as `std::atomic`? This is strictly host code, right?

Good question.  By variables in question , do you mean (lines 133-134):

```
    std::int32_t* const ready_queue = &m_queue[0];
    std::int32_t* const begin_hint  = &m_queue[2 * N];
```

**Comment by PhilMiller** [comment]
Yes

**Comment by ajpowelsnl** [comment]
> Yes

Phil, I'm not sure these variables are host specific, so I'm not sure if it would make sense.

**Comment by dalg24** [comment]
No the functions are annotated host device so you can’t use std::atomic

**Review by dalg24 — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Pull request merged**